src/OsPath/Internal/Include.hs

```diff
@@ -0,0 +1,321 @@
+import System.OsString.Internal.Types (PLATFORM_STRING(..))
+import qualified System.OsString.PLATFORM_NAME as OsString
+
+-- | Path of some base and type.
+--
+-- The type variables are:
+--
+--   * @b@ — base, the base location of the path; absolute or relative.
+--   * @t@ — type, whether file or directory.
+--
+-- Internally it is a PLATFORM_PATH_SINGLE, which can be of two formats only:
+--
+-- 1. File format: @file.txt@, @foo\/bar.txt@, @\/foo\/bar.txt@
+-- 2. Directory format: @foo\/@, @\/foo\/bar\/@
+--
+-- All directories end in a trailing separator. There are no duplicate
+-- path separators @\/\/@, no @..@, no @.\/@, no @~\/@, etc.
+newtype Path b t = Path PLATFORM_PATH
+  deriving (Data, Typeable, Generic)
+
+-- | String equality.
+--
+-- The following property holds:
+--
+-- @show x == show y ≡ x == y@
+instance Eq (Path b t) where
+  (==) (Path x) (Path y) = x == y
+
+-- | String ordering.
+--
+-- The following property holds:
+--
+-- @show x \`compare\` show y ≡ x \`compare\` y@
+instance Ord (Path b t) where
+  compare (Path x) (Path y) = compare x y
+
+-- | Same as 'show . Path.toFilePath'.
+--
+-- The following property holds:
+--
+-- @x == y ≡ show x == show y@
+instance Show (Path b t) where
+  show = show . toFilePath
+
+instance NFData (Path b t) where
+  rnf (Path x) = rnf x
+
+instance ToJSON (Path b t) where
+  toJSON = toJSON . toFilePath
+  {-# INLINE toJSON #-}
+#if MIN_VERSION_aeson(0,10,0)
+  toEncoding = toEncoding . toFilePath
+  {-# INLINE toEncoding #-}
+#endif
+
+instance ToJSONKey (Path b t) where
+  toJSONKey = toJSONKeyText (Text.pack . toFilePath)
+
+instance Hashable (Path b t) where
+  -- A "." is represented as an empty string ("") internally. Hashing ""
+  -- results in a hash that is the same as the salt. To produce a more
+  -- reasonable hash we use "toFilePath" before hashing so that a "" gets
+  -- converted back to a ".".
+  hashWithSalt n path = hashWithSalt n (toFilePath path)
+
+instance forall b t. (Typeable b, Typeable t) => TH.Lift (Path b t) where
+  lift (Path str) = do
+    let b = TH.ConT $ getTCName (Proxy :: Proxy b)
+        t = TH.ConT $ getTCName (Proxy :: Proxy t)
+    [| Path $(TH.lift str) :: Path $(pure b) $(pure t) |]
+    where
+      getTCName :: Typeable a => proxy a -> TH.Name
+      getTCName a = TH.Name occ flav
+        where
+        tc   = typeRepTyCon (typeRep a)
+        occ  = TH.OccName (tyConName tc)
+        flav = TH.NameG TH.TcClsName (TH.PkgName (tyConPackage tc)) (TH.ModName (tyConModule tc))
+
+#if MIN_VERSION_template_haskell(2,17,0)
+  liftTyped = TH.unsafeCodeCoerce . TH.lift
+#elif MIN_VERSION_template_haskell(2,16,0)
+  liftTyped = TH.unsafeTExpCoerce . TH.lift
+#endif
+
+-- | Convert to a 'FilePath' type.
+--
+-- All directories have a trailing slash, so if you want no trailing
+-- slash, you can use 'System.FilePath.dropTrailingPathSeparator' from
+-- the filepath package.
+toFilePath :: Path b t -> FilePath
```

Comment:
IMO, you shouldn't provide this function. There are 3 functions in filepath/or-string to convert from `OsPath` to `FilePath`:

* decodeUtf
* decodeWith
* decodeFS

The decision here to use decodeFS ad-hoc without forcing the user to think about the choice subverts the point of the new API, I think.

Let them get a proper OsPath back and then have them deal with OsPath API.

Reply:
True, I removed the `toFilePath` function.